### PR TITLE
Stable diffusion finetune

### PR DESCRIPTION
**Pull request by A-Jacobson**
Stable diffusion finetuning example

works with stable-diffufusion v1 and 2

includes generation function with negative prompts and classifier free guidance

### CODE REVIEW

File: stable-diffusion/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    <a href="https://join.slack.com/t/mosaicml-community/shared_invite/zt-w0tiddn9-WGTlRpfjcO9J5jyrMub1dg">
+        <img alt="Chat @ Slack" src="https://img.shields.io/badge/slack-chat-2eb67d.svg?logo=slack">
+    </a>
+    <a href="https://github.com/mosaicml/examples/blob/main/LICENSE">
+        <img alt="License" src="https://img.shields.io/badge/License-Apache%202.0-green.svg?logo=slack">
+    </a>
+</p>
+<br />
+
+This folder contains starter code for finetuning Stable Diffusion on the lambda labs pokemon dataset. You can finetune Stable Diffusion v1 or 2, save the model and generate your own pokemon, or swap out the dataset to adapt stable diffusion to your desired domain.
```
Comment (Landanjs) [inline stable-diffusion/README.md:22]: This works with a decent number of HF datasets, right? maybe "fine-tuning Stable Diffusion on image-caption HuggingFace datasets"?

File: stable-diffusion/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+```bash
+git clone mosaicml/examples
+cd examples/stable-diffusion
+```
+## Install requirements
+
+```
+pip install -r requirements.txt 
+```
+## Train the model
```
Comment (Landanjs) [inline stable-diffusion/README.md:37]: Could you provide some details on what will be trained? E.g. maybe stable diffusionv1.4 for 2000 batches on pokemon? I'm not certain what to add, but I imagine it would be nice to know what to expect from running this command

File: stable-diffusion/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+# Results
+Low learning rates and short training times seem to work best. Training longer can quickly lead to overfitting.
+
+TODO: add some nice pictures here
+
+
+
+# Train on different data
+The lambdalabs/pokemon dataset was made with BLIP (a salesforce image captioning model) and uploaded to huggingface datasets. There are many other similar text-to-image datasets available on huggingface here(https://huggingface.co/datasets?task_categories=task_categories:text-to-image). Any of the text-to-image datasets with `image` and `text` columns can be used by simply changing the dataset name in `yamls/finetune.yaml`. Datasets with different formats or color pallets such as spectrograms may require different normalization or pre-processing.
```
Comment (Landanjs) [inline stable-diffusion/README.md:53]:
```suggestion
The lambdalabs/pokemon dataset was made with BLIP (a salesforce image captioning model) and uploaded to huggingface datasets. There are many other similar text-to-image datasets available on huggingface [here](https://huggingface.co/datasets?task_categories=task_categories:text-to-image). Any of the text-to-image datasets with `image` and `text` columns can be used by simply changing the dataset name in `yamls/finetune.yaml`. Datasets with different formats or color pallets such as spectrograms may require different normalization or pre-processing.
```

Comment (Landanjs) [reply stable-diffusion/README.md:53]: Could you also capitalize instances of HuggingFace?

File: stable-diffusion/yamls/finetune.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  eval_batch_size: 32
+  prompts:
+    - "Doge"
+    - "A cute bunny rabbit"
+    - "Yoda"
+    - "An epic landscape photo of a mountain"
+    - "Girl with a pearl earring"
+    - "Cute Obama creature"
+    - "Donald Trump"
+    - "Boris Johnson"
```
Comment (Landanjs) [inline stable-diffusion/yamls/finetune.yaml:27]: I'm a little concerned by political examples (although, I know Lambda Labs used them)

**Review by Landanjs — COMMENTED** [review]
Thanks for getting this together!! On the technical side, everything looks right to me. The biggest thing is improving the format and clarity. A big portion is adjusting the docstrings to be properly formated. Another part is to provide more details on the README. I'm not sure what exactly, but it feels very sparse right now and may be vague to a new user.

**Comment by Landanjs** [comment]
To add to README:
- Overview of files in the directory
- System Recommendations

Also need to update branch with the recent changes to main

File: stable-diffusion/main.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        seed=config.seed)
+    print('Built Trainer\n')
+
+    print('Logging config')
+    log_config(config)
+
+    print('Train!')
+    trainer.fit()
+    # Return trainer for testing purposes
+    return trainer
```
Comment (dblalock) [inline stable-diffusion/main.py:155]: super minor; just trying to shave off a line of code
```suggestion
    return trainer  # Return trainer for testing purposes
```

File: stable-diffusion/model.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                                    max_length=self.tokenizer.model_max_length,
+                                    truncation=True,
+                                    return_tensors="pt")
+        text_embeddings = self.text_encoder(text_input.input_ids.to(device))[0]
+
+        # classifier free guidance + negative prompts
+        # negative prompt is given in place of the unconditional input in classifier free guidance
+        unconditional_input = [
+            ""
+        ] * batch_size if not negative_prompt else negative_prompt
```
Comment (dblalock) [inline stable-diffusion/model.py:172]: mostly trying to get this on one line instead of three
```suggestion
        unconditional_input = negative_prompt or ([""] * batch_size)
```

File: stable-diffusion/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+torchvision>=0.11.0
+mosaicml[all]==0.12.0
+
+
+# --------- loggers --------- #
+wandb
+# neptune-client
+# mlflow
+# comet-ml
+# torch_tb_profiler
```
Comment (dblalock) [inline stable-diffusion/requirements.txt:12]: if these are in here for the user's convenience, we should probably prefix them with a comment to that effect so it doesn't get mistaken for unintentional code

File: stable-diffusion/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# --------- pytorch --------- #
+torch>=1.12.0
+torchvision>=0.11.0
+mosaicml[all]==0.12.0
```
Comment (dblalock) [inline stable-diffusion/requirements.txt:4]:
```suggestion
mosaicml[all]>=0.12.0,<0.13
```
Can we get away with 12.1? trying to get all the examples using the latest version

File: stable-diffusion/yamls/finetune.yaml [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  image_key: "image_tensor"
+  caption_key: "input_ids"
+
+
+optimizer:
+  lr: 1.0e-05
+  weight_decay: 1.0e-2
+
+
+loggers:
```
Comment (dblalock) [inline stable-diffusion/yamls/finetune.yaml:45]:
```suggestion
  caption_key: "input_ids"

optimizer:
  lr: 1.0e-05
  weight_decay: 1.0e-2

loggers:
```
(just making the number of blank lines between sections consistent)

File: stable-diffusion/yamls/mcloud_run.yaml [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    image_key: "image_tensor"
+    caption_key: "input_ids"
+
+
+  optimizer:
+    lr: 2.0e-05
+    weight_decay: 1.0e-3
+
+
+  loggers:
```
Comment (dblalock) [inline stable-diffusion/yamls/mcloud_run.yaml:60]:
```suggestion
    caption_key: "input_ids"

  optimizer:
    lr: 2.0e-05
    weight_decay: 1.0e-3

  loggers:
```
just spacing consistency

**Review by dblalock — COMMENTED** [review]
The only top-level things are that:

1. The repo got restructured so this needs to be a subdirectory of the `examples/` package.
2. This example needs to get added to the installation, tests, and CI.

(2) entails:

- Modifying `setup.py`'s [list of examples](https://github.com/mosaicml/examples/blob/main/setup.py#L12)
- Modifying the `lint` and `test` targets [in the Makefile](https://github.com/mosaicml/examples/blob/main/Makefile#L35)
- Modifying the lists of examples in the [`.github/workflows`](https://github.com/mosaicml/examples/tree/main/.github/workflows) yamls.

Or, alternatively, making all these reference some central list of subdirectories, which would be even nicer.

In the mean time, running `make style` in the project root should fix a lot of the style stuff I was initially commenting on.

File: stable-diffusion/README.md (hunk 1 above)
Comment (A-Jacobson) [reply stable-diffusion/README.md:22]: agreed, i started adding in the other datasets after the fact.

File: stable-diffusion/yamls/finetune.yaml (hunk 4 above)
Comment (A-Jacobson) [reply stable-diffusion/yamls/finetune.yaml:27]: good point. I don't love them either. Had them in to validate our stuff was working. no need to keep them around.

File: stable-diffusion/data.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        [example["image_tensor"] for example in examples])
+    image_tensor = image_tensor.to(
+        memory_format=torch.contiguous_format).float()
+    input_ids = torch.stack([example["input_ids"] for example in examples])
+    return {"image_tensor": image_tensor, "input_ids": input_ids}
+
+
+def build_image_caption_datapsec(name:str,
+                                resolution: int,
+                                tokenizer: callable,
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/data.py:31]: it's a builtin type like int/string. just means it's a function.

File: stable-diffusion/main.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    # Initialize dist to ensure dataset is only downloaded by rank 0
+    device = 'gpu' if torch.cuda.is_available() else 'cpu'
+    if dist.get_world_size() > 1:
+        dist.initialize_dist(device, timeout=180)
+        train_batch_size //= dist.get_world_size()
+        train_batch_size = train_batch_size or 1
+
+        eval_batch_size //= dist.get_world_size()
+        eval_batch_size = eval_batch_size or 1
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/main.py:60]: i guess we could? the other scripts will error if the batch size is less than world size too now that i think about it.

[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/main.py:60]: could throw an error. now that i think about it the other scripts will likely error on similar settings as well.

File: stable-diffusion/model.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+        if negative_prompt:
+            negative_prompt_bs = 1 if isinstance(negative_prompt,
+                                                 str) else len(negative_prompt)
+            if negative_prompt_bs != batch_size:
+                raise ValueError(
+                    f'len(prompts) and len(negative_prompts) must be the same. A negative prompt must be provided for each given prompt.'
+                )
+
+        vae_scale_factor = 8
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/model.py:154]: not one that i could find in it's config =(

File: stable-diffusion/model.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+                         conditioning)['sample'], target
+
+    def loss(self, outputs, batch):
+        """loss between unet output and added noise, typically mse"""
+        return self.loss_fn(outputs[0], outputs[1])
+
+    def eval_forward(self, batch, outputs=None):
+        if outputs:
+            return outputs
+        return self.generate(batch)
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/model.py:122]: I originally had it setup that way but any evaluation metrics are going to be run on generated images so i believe this makes more sense. outputs shouldn't be none, it was just boilerplate from the base composermodel.

File: stable-diffusion/data.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    dataset = PromptDataset(prompts)
+    sampler = dist.get_sampler(dataset, drop_last=False, shuffle=False)
+    
+    def split_list(l, microbatch_size: int): # needed for gradient accumulation
+        if len(l) < microbatch_size:
+            microbatch_size = len(l)
+        num_microbatches = math.ceil(len(l) / microbatch_size)
+        # Note: this is to match the behavior of tensor.chunk, which is used in _split_tensor
+        chunked_microbatch_size = math.ceil(len(l) / num_microbatches)
+        return [l[start:start + chunked_microbatch_size] for start in range(0, len(l), chunked_microbatch_size)]
```
Comment (Landanjs) [inline stable-diffusion/data.py:110]: Could this be avoided if `generate` took in tokenized text instead of raw strings? The change seems worth it to avoid this extra function

Comment (A-Jacobson) [reply stable-diffusion/data.py:110]: Composer is/was being updated to handle strings. I'll test the new version and update accordingly. Generate taking tokens would be clunky to use at inference time imo.

File: stable-diffusion/main.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+                                          batch_size=eval_batch_size)
+
+    # Optimizer
+    print('Building optimizer and learning rate scheduler')
+
+    optimizer = torch.optim.AdamW(params=model.parameters(),
+                                  lr=config.optimizer.lr,
+                                  weight_decay=config.optimizer.weight_decay)
+
+    # Learning rate scheduler: LR warmup for 8 epochs, then cosine decay for the rest of training
```
[the remark this answers is not in the document]
Comment (tonyf) [reply stable-diffusion/main.py:94]: Where is warmup implemented?

File: stable-diffusion/data.py (hunk 15 above)
Comment (Landanjs) [reply stable-diffusion/data.py:110]: For inference, it shouldn't be too bad to add a tokenizer as a preprocessor to input data, right? 

I was motivated to ask since it was taking me a few hours to get COCO captions dataloader to output raw strings and for composer to accept them. It made it way easier when `generate()` could accept tokenized text.

File: stable-diffusion/main.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply stable-diffusion/main.py]: It's just an erroneous comment. Sorry to mislead you!

File: examples/stable_diffusion/model.py [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+            num_inference_steps (int): The number of denoising steps. 
+                More denoising steps usually lead to a higher quality image at the expense 
+                of slower inference. Default: `50`.
+            guidance_scale (float): Guidance scale as defined in 
+                Classifier-Free Diffusion Guidance. guidance_scale is defined as w of equation 
+                2. of Imagen Paper. Guidance scale is enabled by setting guidance_scale > 1. 
+                Higher guidance scale encourages to generate images that are closely linked 
+                to the text prompt, usually at the expense of lower image quality. 
+                Default: `7.5`.
+            negative_prompt (str or List[str]) — The prompt or prompts not to guide the image generation. Ignored when not using guidance (i.e., ignored if guidance_scale is less than 1). Must be the same length as list of prompts.
```
Comment (Landanjs) [inline examples/stable_diffusion/model.py:165]: Cap line length to 80

File: examples/stable_diffusion/model.py [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+            metrics_dict = {}
+            for name, metric in metrics.items():
+                assert isinstance(metric, Metric)
+                metrics_dict[name] = metric
+        return metrics_dict
+
+    def update_metric(self, batch, outputs, metric):
+        prompts = batch
+        if metric.__class__.__name__ == 'CLIPScore':
+            metric.update(outputs, prompts)
```
Comment (Landanjs) [inline examples/stable_diffusion/model.py:270]: Did you get CLIPScore to work?

File: examples/stable_diffusion/tests/test_data.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+    dataloader = DataLoader(
+        dataset=dataset,
+        batch_size=batch_size,
+        sampler=sampler,
+        drop_last=drop_last,
+        **dataloader_kwargs,
+    )
+
+    return dataloader
```
Comment (Landanjs) [inline examples/stable_diffusion/tests/test_data.py:54]: Add endline

File: examples/stable_diffusion/README.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+# Results
+The default config is `yamls/finetune.yaml`. It trains only the Unet portion of `Stable Diffusion v1-4` for `2000 batches` on the lambda labs `Pokemon` dataset. A default set of evaluation prompts are included in the yaml and will be logged to weights & baises every `200 batches`. 
+
+Results from the default config and the prompt  `"a magestic shiba inu doge wearing a blue sweater"`:
+<p align="center">
+  <picture>
+    <img alt="model outputs of a Shiba inu pokemon in a blue sweater" src="./assets/A_Magestic_Shiba_Inu_Doge_wearing_a_blue_sweater_2000_batches.png">
+  </picture>
+</p>
```
Comment (Landanjs) [inline examples/stable_diffusion/README.md:50]: Could you also add a set of images to the top of the README?

**Review by Landanjs — APPROVED** [review]
LGTM! Great job on integrating the comments!

The main comment is to add a set of generated images to the top of the README. Otherwise, just style comments.

File: examples/stable_diffusion/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,20 @@
+# --------- pytorch --------- #
+torch>=1.12.0
+torchvision>=0.11.0
+mosaicml[all]>=0.12.1
+
+
+# --------- loggers --------- #
+wandb
```
Comment (dakinggg) [inline examples/stable_diffusion/requirements.txt:8]: suggest either pinning or remove (its included as part of `mosaicml[all]`
